RavenDB-14555 - Removing thread local usages

ayende opened the pull request.

ayende committed:
RavenDB-14555 Replace CloseableThreadLocal with a lightweight implementation

ayende committed:
RavenDB-14555 Remove ThreadLocal usage in favor of passing the value using the context

ayende committed:
RavenDB-14555 Remove ThreadLocal, we can use a simple ThreadStatic here, instead

ayende committed:
RavenDB-14555 Introducing LightThreadLocal impl instead of the default expensive one

karmeli87 commented:
I think we also want to remove the usage of `ThreadLocal` in the client.
Two can be found here:
https://github.com/ravendb/ravendb/blob/af3b7efe4e16d961ed146b386da0569867da9c12/src/Raven.Client/Documents/Conventions/JsonNetBlittableEntitySerializer.cs#L13

ppekrol commented on src/Sparrow/Utils/LightThreadLocal.cs:
```diff
@@ -0,0 +1,97 @@
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Sparrow.Utils
```
can we move it to Sparrow.Threading? Looks more appropriate?

ppekrol commented on src/Voron/Util/ThreadHoppingReaderWriterLock.cs:
```diff
@@ -36,7 +32,7 @@ public void EnterWriteLock()
 
             // try take ownership on lock
-            var currentLock = Interlocked.CompareExchange(ref _writeLockOwnerThreadId, _localThreadIdCache.Value, 0);
+            var currentLock = Interlocked.CompareExchange(ref _writeLockOwnerThreadId, Thread.CurrentThread.ManagedThreadId, 0);
```
Wasn't the cache crated because accessing 'Thread.CurrentThread.ManagedThreadId' is expensive? Maybe it is better to access 'NativeMemory.CurrentThreadStats.ManagedThreadId' instead?

Anyway, at least we can create a variable in this method to avoid accessing it twice.

ayende replied:
It used to be expensive, but now, I don't think so.
See:

https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs#L19

https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs#L156

https://github.com/dotnet/runtime/blob/master/src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs#L161

https://github.com/dotnet/runtime/blob/fcd862e06413a000f9cafa9d2f359226c60b9b42/src/coreclr/src/vm/object.h#L1426

So this ends up being as fast as our own impl

ayende commented:
> I think we also want to remove the usage of `ThreadLocal` in the client.
> Two can be found here:
> https://github.com/ravendb/ravendb/blob/af3b7efe4e16d961ed146b386da0569867da9c12/src/Raven.Client/Documents/Conventions/JsonNetBlittableEntitySerializer.cs#L13

Those are probably going to be only twice per DocumentStore, which means that they are likely _very_ small. 
I want to first see that everything is fine with the actual impl

ayende committed:
RavenDB-14555 Call ManagedThreadId once

ayende committed:
RavenDB-14555 Ensure that we don't hold on to the thread local instance for thread statics, implement finalizer

ayende committed:
RavenDB-14555 Move to better namespace, use LightThreadLocal on client as well, remove all usages of sysytem's ThreadLocal<T> from the codebase

ppekrol committed:
RavenDB-14555 prevent double dispose

ppekrol commented:
Rebased and used newest Lucene

ayende committed:
RavenDB-14555
- Avoid double dispose without locking (called from finalizer, want to avoid it)
- Making classes sealed
- Rearranged classes

ppekrol committed:
RavenDB-14555 added Lucene.Net.3.0.22.nupkg

ayende committed:
RavenDB-14555 Handle cleanup of local thread set when the Threading/LightWeightThreadLocal instance is disposed

ayende committed:
RavenDB-14555 Updating Lucene for the new thread local
